crypt-tune-costs.c

```diff
@@ -0,0 +1,740 @@
+{
+  if (method->usage == HU_LEGACY || method->usage == HU_DISABLED)
+    {
+      if (verbosity >= 1)
+        fprintf (stderr, "# %s: %s, skipping\n",
+                 method->name, usage_keyword[method->usage]);
+      return;
+    }
+  if (method->hct == HCT_FIXED)
+    {
+      if (verbosity >= 1)
+        fprintf (stderr, "# %s: fixed cost\n", method->name);
+      method->nrounds = 0;
+      method->elapsed = time_crypt (method, 0);
+    }
+  else if (method->hct == HCT_LINEAR)
+    {
+      if (verbosity >= 1)
+        fprintf (stderr, "# %s: linear cost\n", method->name);
+      tune_linear_cost (method, elapsed_target);
+    }
+  else /* HCT_EXPON */
+    {
+      if (verbosity >= 1)
+        fprintf (stderr, "# %s: exponential cost\n", method->name);
+      tune_expon_cost (method, elapsed_target);
+    }
+
+  if (method->elapsed < elapsed_target)
+    method->usage = strict ? HU_DISABLED : HU_LEGACY;
+}
+
+static void
+tune_all_costs (double elapsed_target, bool strict)
+{
+  size_t i;
+  for (i = 0; i < ARRAY_SIZE (hash_methods); i++)
+    tune_cost (&hash_methods[i], elapsed_target, strict);
+
+  /* Make sure that at least one hashing method is still enabled, and
+     one hashing method is preferred.  */
+  static_assert (ARRAY_SIZE (hash_methods) <= INT_MAX,
+                 "too many hash methods for 'int'");
+  int first_enabled = -1;
+
+  for (i = 0; i < ARRAY_SIZE (hash_methods); i++)
+    if (hash_methods[i].usage == HU_PREFER)
+      return;
+    else if (hash_methods[i].usage == HU_ENABLED && first_enabled == -1)
+      first_enabled = (int)i;
+
+  /* If we get here, no method was preferred.
+     Are there any enabled methods at all?  */
+  if (first_enabled == -1)
+    {
+      fprintf (stderr, "%s: no enabled hashing method can take %6.2fms\n",
+               program_name, elapsed_target);
+      exit (1);
+    }
+
+  /* The hash_methods table is in descending order of cryptographic
+     strength, so when no explicit selection was made, use the first
+     method that's enabled as the preferred method.  */
+  hash_methods[first_enabled].usage = HU_PREFER;
+}
+
+static void
+write_config (void)
+{
+  fputs ("# crypt.conf generated by crypt-tune-costs.\n"
+         "# Rounds settings are tuned for this computer.\n\n",
+         stdout);
+  for (size_t i = 0; i < ARRAY_SIZE (hash_methods); i++)
+    if (hash_methods[i].nrounds != 0)
+      printf ("%-12s%-12srounds=%u\t# %6.2fms\n",
+              hash_methods[i].name,
+              usage_keyword[hash_methods[i].usage],
+              hash_methods[i].nrounds,
+              hash_methods[i].elapsed);
+    else
+      printf ("%-12s%s\n",
+              hash_methods[i].name,
+              usage_keyword[hash_methods[i].usage]);
+
+  if (fflush (stdout) || fclose (stdout))
+    {
+      fprintf (stderr, "%s: stdout: %s\n", program_name, strerror (errno));
+      exit (1);
+    }
+}
+
+static NORETURN
```

Comment:
You should use the [`format` attribute](http://clang.llvm.org/docs/AttributeReference.html#format) here, so Clang doesn't complain about `-Wformat-nonliteral`.  Using `__attribute__((__format__ (__printf__, 1, 0)))` should be sane according to the [GCC online docs](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes).

Reply:
Resolved in https://github.com/besser82/libxcrypt/pull/26/commits/4953f92c0db7eb45d83dd293635bf12f70799839.